File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        Queue<int[]> q = new ArrayDeque<>();
        int directions[][] = {{1, 0}, {-1, 0}, {0,1}, {0,-1}};

        int ans = 0;

        for(int i =0; i< m; i++)
        {
            for(int j = 0; j< n; j++)
            {
                if(grid[i][j] == '1')
                {
                    ans++;
                    q.offer(new int[]{i,j});

                    while(!q.isEmpty())
                    {
                        int curr[] = q.poll();
                        int x = curr[0];
                        int y = curr[1];

                        if(x<0 || x>=m || y<0 || y>=n || grid[x][y] != '1')
                            continue;

                        grid[x][y] = '0';

                        for(int dir[]: directions)
                        {
                            int nx = x + dir[0];
                            int ny = y + dir[1];
                            if(nx>=0 && nx < m && ny >=0 && ny <n && grid[nx][ny]=='1')
                            {
                                q.offer(new int[]{nx, ny});
                            }
                        }
                    }

                }
            }
        }
        return ans;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        ans = 0

        for i in range(0, m):
            for j in range(0, n):
                if grid[i][j] == "1":
                    ans += 1
                    self.dfs(grid, i, j, m, n)
        return ans

    def dfs(self, grid, i, j, m, n):
        if i<0 or i>=m or j < 0 or j >= n or grid[i][j] !="1":
            return
        
        grid[i][j] = 0
        self.dfs(grid, i+1, j, m, n)
        self.dfs(grid, i-1, j, m, n)
        self.dfs(grid, i, j+1, m, n)
        self.dfs(grid, i, j-1, m, n)
        
